site/app/models/NotificationFactory.php

```diff
@@ -0,0 +1,126 @@
+        $notification_recipients = array_unique(array_merge($parent_authors, $users_with_notification_preference, $thread_authors_notification_preference));
+        $this->createAndSendNotifications($event, $notification_recipients);
+
+        $users_with_email_preference = $this->core->getQueries()->getAllUsersWithPreference("all_new_posts_email");
+        $thread_authors_email_preference = $this->core->getQueries()->getAllThreadAuthors($thread_id,"reply_in_post_thread_email");
+        $email_recipients = array_unique(array_merge($parent_authors, $users_with_email_preference, $thread_authors_email_preference));
+        $this->createAndSendEmails($event, $email_recipients);
+
+    }
+
+    /**
+     * handles the event of a post deleted, undeleted, edited and merged
+     * @param array $event
+     */
+    public function onPostModified(array $event) {
+        //
+        $notification_recipients = $this->core->getQueries()->getAllUsersWithPreference($event['preference']);
+        $notification_recipients[] = $event['recipient'];
+        $notification_recipients = array_unique($notification_recipients);
+        $this->createAndSendNotifications($event,$notification_recipients);
+
+        $email_recipients =  $this->core->getQueries()->getAllUsersWithPreference($event['preference'].'_email');
+        $email_recipients[] = $event['recipient'];
+        $email_recipients = array_unique($email_recipients);
+        $this->createAndSendEmails($event,$email_recipients);
+    }
+
+    // ***********************************GRADE INQUIRY NOTIFICATIONS***********************************
+
+    /**
+     * @param array $event
+     */
+    public function onGradeInquiryEvent(array $event) {
+        // TODO::Allow users to have preference on grade inquiry events
+        $recipients = $event['recipients'];
+        $this->createAndSendNotifications($event,$recipients);
+        $this->createAndSendEmails($event,$recipients);
+    }
+
+    // ***********************************SENDERS***********************************
+    /**
+     * @param array $event
+     * @param array $recipients
+     */
+    public function createAndSendNotifications(array $event, array $recipients) {
+        // if there are no recipients return
+        if (empty($recipients)) {
+            return;
+        }
+        $event['user_id'] = $this->core->getUser()->getId();
+        $notification = Notification::createNotification($this->core,$event);
+        $this->core->getQueries()->pushNotifications($notification,$recipients);
+    }
+
+    /**
+     * @param array $event
+     * @param array $recipients
+     */
+    public function createAndSendEmails(array $event, array $recipients) {
+        if (empty($recipients)) {
+            return;
+        }
+        $event["user_id"] = $this->core->getUser()->getId();
+        $email = new Email($this->core,$event);
+        $this->core->getQueries()->pushEmails($email,$recipients);
```

Comment:
The `user_id` of emails is who should receive the email, not who initiated an action that sends emails.

Reply:
Should the email object resemble a row entry in the emails table? I am using the object a contianer for what the email is and the list of recipients? Can I just rename user_id to author. I think that is definitely clearer.

Reply:
For rows in the email table, user_id stands for who we are sending the email to. We don't store the sender in the table. @bmcutler would have to answer if she thinks tracking that information would be useful, though I'm not sure what it would be used for. If it is something we want, probably requires some sort of answer of "what information are we interested in and the potential questions we might ask about it" when looking through the logs.

Reply:
We didn't briefly consider the possibility of having multiple sender emails (if available) -- but that's not the current design/implementation.  The actual sender is the same for every email.  No need to store that.

In some cases we do have a real user who caused the email to be sent.  E.g. an announcement.  And we would like (in some cases -- perhaps up to the user?) to include their name in the body of the message.  But that's not a separate column of the table.